hypothesis-python/src/hypothesis/strategies/_internal/numbers.py

```diff
@@ -265,9 +279,41 @@ def do_draw(self, data):
         if self.width < 64:
             f = float_of(f, self.width)
         assume(self.lower_bound <= f <= self.upper_bound)
+        if not self.allow_subnormal:
+            assume(f == 0 or abs(f) > width_smallest_normals[self.width])
         return f
 
 
+def safe_next_down(value, width, allow_subnormal):
+    smallest_normal = width_smallest_normals[width]
+    if allow_subnormal or value <= -smallest_normal or value > smallest_normal:
+        return next_down(value, width)
+    else:
+        if 0.0 < value <= smallest_normal:
+            return 0.0
+        else:
+            return -smallest_normal
```

Comment:
I think yours mis-handles negative zero, so let's just get the next_down and skip past subnormals if we need to:

```suggestion
def next_down_normal(value, width, allow_subnormal):
    value = next_down(value, width)
    if (not allow_subnormal) and 0 < abs(value) < width_smallest_normals[width]:
        return 0.0 if value > 0 else -width_smallest_normals[width]
    return value
```